bitvm/src/signatures/public.rs

```diff
@@ -0,0 +1,838 @@
+/// Winternitz signatures for 64-byte messages.
+pub struct Wots64;
+/// Winternitz signatures for 72-byte messages.
+pub struct Wots72;
+/// Winternitz signatures for 80-byte messages.
+pub struct Wots80;
+
+macro_rules! impl_wots {
+    ($name:ident, $msg_byte_len:expr, $converter:ty) => {
+        impl Wots for $name {
+            type Converter = $converter;
+            /// The public key type for this Winternitz signing algorithm.
+            type PublicKey = [[u8; 20]; Self::TOTAL_DIGIT_LEN as usize];
+            /// The message type for this Winternitz signing algorithm.
+            ///
+            /// All messages have the same fixed length.
+            type Message = [u8; Self::MSG_BYTE_LEN as usize];
+            /// The signature type of this Winternitz signing algorithm.
+            type Signature = [[u8; 21]; Self::TOTAL_DIGIT_LEN as usize];
+
+            /// The number of bytes in a message.
+            const MSG_BYTE_LEN: u32 = $msg_byte_len;
+        }
+
+        impl CompactWots for $name {
+            /// The compact signature type of this Winternitz signing algorithm.
+            type CompactSignature = [[u8; 20]; Self::TOTAL_DIGIT_LEN as usize];
+        }
+    };
+}
+
+impl_wots!(Wots4, 4, VoidConverter);
+impl_wots!(Wots16, 16, VoidConverter);
+impl_wots!(Wots32, 32, VoidConverter);
+impl_wots!(Wots64, 64, VoidConverter);
+impl_wots!(Wots72, 72, ToBytesConverter);
+impl_wots!(Wots80, 80, VoidConverter);
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::u32::u32_std::u32_compress;
+    use crate::{bn254::g1::G1Affine, execute_script, signatures::utils};
+    use crate::{execute_script_with_inputs, ExecuteInfo};
+
+    use std::convert::TryFrom;
+    use std::fs::File;
+    use std::io;
+    use std::io::Read;
+    #[allow(unused_imports)]
+    use std::io::Write; // needed to generate test vectors (see below)
+
+    use ark_ff::UniformRand as _;
+    use ark_std::test_rng;
+    use bitcoin::hex::{DisplayHex, FromHex};
+    use bitcoin::script::read_scriptint;
+    use bitcoin_script::script;
+    use rand::{RngCore as _, SeedableRng as _};
+    use serde::ser::SerializeSeq;
+    use serde::{Deserialize, Deserializer, Serialize, Serializer};
+    use serde_json;
+
+    fn extract_witness_from_stack(res: ExecuteInfo) -> Vec<Vec<u8>> {
+        res.final_stack.0.iter_str().fold(vec![], |mut vector, x| {
+            vector.push(x);
+            vector
+        })
+    }
+
+    /// Compare two elements of n length.
+    /// If them are not equal, return script's failure directly.
+    pub fn equalverify(n: usize) -> Script {
+        script!(
+            for _ in 0..n {
+                OP_TOALTSTACK
+            }
+
+            for i in 1..n {
+                {i}
+                OP_ROLL
+            }
+
+            for _ in 0..n {
+                OP_FROMALTSTACK
+                OP_EQUALVERIFY
+            }
+        )
+    }
+
+    pub fn u32_witness_to_bytes(witness: Vec<Vec<u8>>) -> Vec<u8> {
```

Comment:
are we using this fn outside this test module?
if not, make this non`pub`

Reply:
Also please include at least brief internal documentation for this.